### ISSUE TITLE
list containsExactlyInAnyOrderElementsOf with usingRecursiveFieldByFieldElementComparator does not honor Optionals

### ISSUE THREAD

**Issue by Brummolix** [joel-costigliola/assertj-core#1726]
#### Summary
When you combine containsExactlyInAnyOrderElementsOf  with usingRecursiveFieldByFieldElementComparator  and also have getters with optionals, they will not be honored

#### Example

```java
package assertj;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import java.util.Optional;

import org.junit.jupiter.api.Test;

public class TestClass {

	public static class Class1 {

		private final String value;

		public Class1(final String value) {
			this.value = value;
		}

		public String getValue() {
			return value;
		}

		@Override
		public String toString() {
			return "Class1 [value=" + value + "]";
		}
	}

	public static class Class2WithOptional {

		private final Class1 value;

		public Class2WithOptional(final Class1 value) {
			this.value = value;
		}

		public Optional<Class1> getValue() {
			return Optional.ofNullable(value);
		}

		@Override
		public String toString() {
			return "Class2WithOptional [value=" + value + "]";
		}
	}

	@Test
	void testRecursiveEquals() {
		Class2WithOptional testClass1 = new Class2WithOptional(new Class1("value"));
		Class2WithOptional testClass2 = new Class2WithOptional(new Class1("value"));

		assertThat(testClass1)
				.usingRecursiveComparison()
				.isEqualTo(testClass2);
	}

	@Test
	void testRecursiveEqualsOnList() {
		Class2WithOptional testClass1 = new Class2WithOptional(new Class1("value"));
		Class2WithOptional testClass2 = new Class2WithOptional(new Class1("value"));

		var list1 = List.of(testClass1);
		var list2 = List.of(testClass2);

		assertThat(list1)
				.usingRecursiveFieldByFieldElementComparator()
				.containsExactlyInAnyOrderElementsOf(list2);
	}
}

```

**Comment by joel-costigliola** [comment]
@Brummolix which version are you using? I'm wondering if this is related to https://github.com/joel-costigliola/assertj-core/issues/1424.

**Comment by Brummolix** [comment]
testRecursiveEqualsOnList fails in 3.13.2 and 3.14.0 (not tested with older versions)

**Comment by joel-costigliola** [comment]
thanks for checking, we will look into it!

**Comment by joel-costigliola** [comment]
`usingRecursiveFieldByFieldElementComparator` is not as advanced as `usingRecursiveComparison` in term of recursive comparison logic, the latter comparing correctly optionals.

We are going to deprecate `usingRecursiveFieldByFieldElementComparator` in favor `usingRecursiveComparison` when https://github.com/joel-costigliola/assertj-core/issues/1238 is going to be addressed (hopefully is 3.15.0 but that's not a straightforward piece of work so not 100% sure of that).

On the meantime there is a workaround for `testRecursiveEqualsOnList` if is `isEqualTo` is sufficient for you, simply cast the list to an object and call  `usingRecursiveComparison`:
```java 
    assertThat((Object) list1).usingRecursiveComparison()
                              .isEqualTo(list2);
```

**Comment by joel-costigliola** [comment]
Closing as addressed by #1238
